nexus/src/db/datastore.rs

```diff
@@ -3394,6 +3396,93 @@ impl DataStore {
                 ))
             })
     }
+
+    pub async fn allocate_static_v6_address(
+        &self,
+        allocation_config: &StaticV6AllocationConfig,
+    ) -> Result<Ipv6Addr, Error> {
+        use db::schema::static_v6_address::dsl;
+
+        let allocation_config = allocation_config.clone();
+
+        self.pool()
+            .transaction(move |conn| {
+                // First, search for released addresses
+                let result: Result<StaticV6Address, diesel::result::Error> =
+                    dsl::static_v6_address
+                        .filter(dsl::allocated.eq(false))
+                        .limit(1)
+                        .get_result(conn);
+
+                if let Ok(row) = result {
+                    diesel::update(dsl::static_v6_address)
+                        .filter(dsl::address.eq(&row.address))
+                        .set(dsl::allocated.eq(true))
+                        .execute(conn)?;
+
+                    return Ok(row.address.parse().unwrap());
+                }
+
+                // If no released addresses exist, allocate a new one
+                // XXX there's got to be a better way! store the iterator
+                // somewhere?
+                for ip in allocation_config.subnet.iter() {
+                    // Do not allocate less than the minimum address
```

Comment:
Nit: We could probably just start at the minimum address, rather than looping through all the non-allocatable addresses?

Reply:
this is also gone now